processor/spanmetricsprocessor/processor.go

```diff
@@ -231,21 +245,40 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) e
 // writes the raw metrics data into the metrics object.
 func (p *processorImp) buildMetrics() *pdata.Metrics {
 	m := pdata.NewMetrics()
-	ilm := m.ResourceMetrics().AppendEmpty().InstrumentationLibraryMetrics().AppendEmpty()
-	ilm.InstrumentationLibrary().SetName("spanmetricsprocessor")
+	rms := m.ResourceMetrics()
+	for key := range p.resourceAttrList {
+		p.lock.RLock()
+		resourceAttrKey := resourceKey(key)
+		resourceAttributesMap := p.resourceKeyToDimensions[resourceAttrKey]
+
+		// if the service name doesn't exist, we treat it as invalid and do not generate a trace
+		if _, ok := resourceAttributesMap.Get(serviceNameKey); !ok {
+			continue
+		}
+
+		rm := rms.AppendEmpty()
+
+		resourceAttributesMap.Range(func(k string, v pdata.AttributeValue) bool {
+			rm.Resource().Attributes().Insert(k, v)
+			return true
+		})
 
-	p.lock.RLock()
-	p.collectCallMetrics(ilm)
-	p.collectLatencyMetrics(ilm)
-	p.lock.RUnlock()
+		ilm := rm.InstrumentationLibraryMetrics().AppendEmpty()
+		ilm.InstrumentationLibrary().SetName("spanmetricsprocessor")
+
+		// build metrics per resource
+		p.collectCallMetrics(ilm, resourceAttrKey)
```

Comment:
I feel like we can (probably?) create new metrics under new resources without complicating the maps with `resourceAttrKey`. Happy to be shown otherwise. 

But can't we iterate through the resources and still maintain the same metric keys? They will still be unique and include the resource attributes.

It's just quite a big change and if you want this to be merged upstream then I have worries if it's big and complex, there could be some push back.

Reply:
Yeah, we could in theory create a new resource for every metric even if they have the same resource attributes but this way is more optimised for the processor in terms of memory. It's a trade-off I guess, simplicity of code vs memory. Personally I think memory optimisation is more important and I don't think the changes are that complex (albeit definitely more complex than the original). Do you think it's more worth it to have simple code?